stdlib/universe/limit.go

```diff
@@ -238,3 +246,112 @@ func (t *limitTransformation) UpdateProcessingTime(id execute.DatasetID, pt exec
 func (t *limitTransformation) Finish(id execute.DatasetID, err error) {
 	t.d.Finish(err)
 }
+
+type limitState struct {
+	n      int
+	offset int
+}
+type limitTransformationAdapter struct {
+	limitTransformation *limitTransformation
+}
+
+func (*limitTransformationAdapter) Close() error {
+	return nil
+}
+
+func (t *limitTransformationAdapter) Process(
+	chunk table.Chunk,
+	state interface{},
+	dataset *execute.TransportDataset,
+	_ arrowmem.Allocator,
+) (interface{}, bool, error) {
+
+	var state_ *limitState
+	// `.Process` is reentrant, so to speak. The first invocation will not
+	// include a value for `state`. Initialization happens here then is passed
+	// in/out for the subsequent calls.
+	if state == nil {
+		state_ = &limitState{n: t.limitTransformation.n, offset: t.limitTransformation.offset}
+	} else {
+		state_ = state.(*limitState)
+	}
+	return t.processChunk(chunk, state_, dataset)
+}
+
+func (t *limitTransformationAdapter) processChunk(
+	chunk table.Chunk,
+	state *limitState,
+	dataset *execute.TransportDataset,
+) (*limitState, bool, error) {
+	chunkLen := chunk.Len()
+
+	// Pass empty chunks along to downstream transformations.
+	if chunkLen == 0 {
+		chunk.Retain()
+		if err := dataset.Process(chunk); err != nil {
+			return nil, false, err
+		}
+		return state, true, nil
+	}
+
+	if chunkLen <= state.offset {
+		state.offset -= chunkLen
+		return state, true, nil
+	}
+
+	start := state.offset
+	stop := chunkLen
+	count := stop - start
+	if count > state.n {
+		count = state.n
+		stop = start + count
+	}
+
+	// When n=0, clamp the offsets such that we emit an empty chunk.
+	if state.n <= 0 {
+		start = 0
```

Comment:
This seems like overwriting the variable(start, stop, count) in case of n<=0, Could you please put this corner case at the top.

something like - 
```
if n <= 0 {
   return state, true, nil
}
```

Reply:
I can try this again, but I struggled with this much of the day yesterday.

For whatever reason, the tests fail if we don't produce an "empty" chunk, and when I tried to do this directly I kept seeing nil pointer panics. As a last ditch, I set all the offsets to 0 and let the main block where we produce a chunk at the bottom handle this for me.

Reply:
I move this up to be handled along with the zero length chunk case at the top, but it did require me to duplicate much of the code later in the method (ref: https://github.com/influxdata/flux/pull/4514/commits/1cbf0a42873f2e1696f58a662528da755ce8a36b).

@skartikey do you know of a more terse approach for building an empty chunk? I assume the work here is essentially just to ensure we're passing along the schema information.

Reply:
I guess I'll merge as-is and tighten this up down the road as needed.